sdk/keyvault/internal/challenge_policy_test.go

```diff
@@ -136,15 +144,62 @@ func TestFindScopeAndTenant(t *testing.T) {
 	}
 
 	resp.Header.Set("WWW-Authenticate", "this is an invalid value")
-	err = p.findScopeAndTenant(&resp)
+	err = p.findScopeAndTenant(&resp, req)
 	var challengeError *challengePolicyError
 	require.ErrorAs(t, err, &challengeError)
 
 	resp.Header = http.Header{}
-	err = p.findScopeAndTenant(&resp)
+	err = p.findScopeAndTenant(&resp, req)
 	require.ErrorAs(t, err, &challengeError)
 }
 
+func TestResourceVerification(t *testing.T) {
+	for _, test := range []struct {
+		challenge, resource string
+		err                 bool
+	}{
+		{authResource, "https://vault.azure.net", false},
+		{authScope, "https://vault.azure.net/.default", false},
+
+		// error cases: resource/scope doesn't match the requested vault's host (vault.azure.net)
+		{authResource, "https://vault.azure.cn", true},
+		{authResource, "https://myvault.azure.net", true},
+		{authScope, "https://vault.azure.cn/.default", true},
+		{authScope, "https://myvault.azure.net/.default", true},
+	} {
+		t.Run("", func(t *testing.T) {
+			srv, close := mock.NewServer(mock.WithTransformAllRequestsToTestServerUrl())
+			defer close()
+			srv.AppendResponse(
+				mock.WithHeader("WWW-Authenticate", fmt.Sprintf(test.challenge, fakeTenant, test.resource)),
+				mock.WithStatusCode(401),
+			)
+			srv.AppendResponse()
+
+			cred := credentialFunc(func(context.Context, policy.TokenRequestOptions) (azcore.AccessToken, error) {
+				return azcore.AccessToken{Token: "***", ExpiresOn: time.Now().Add(time.Hour)}, nil
+			})
+			p := NewKeyVaultChallengePolicy(cred)
+			pl := runtime.NewPipeline("", "",
+				runtime.PipelineOptions{PerCall: []policy.Policy{p}},
+				&policy.ClientOptions{Transport: srv},
+			)
+			req, err := runtime.NewRequest(context.Background(), "GET", "https://42.vault.azure.net")
+			require.NoError(t, err)
+			_, err = pl.Do(req)
+			if test.err {
+				expected := fmt.Sprintf(`the challenge resource "%s" doesn't match the requested domain`, test.resource)
+				require.EqualError(t, err, expected)
+				if _, ok := err.(*challengePolicyError); !ok {
```

Comment:
There's a require for this too if you want to use it:

`require.ErrorAs()`

Reply:
I used the type assert because I want this test to fail if the returned error only wraps `challengePolicyError`.